process/block/preprocess/miniBlocksCompaction.go

```diff
@@ -299,39 +283,74 @@ func (mbc *miniBlocksCompaction) computeMerge(
 	srcIndex int,
 ) bool {
 
-	canBeFullyMerged := mbc.mapMiniBlockGasConsumed[dstIndex]+mbc.mapMiniBlockGasConsumed[srcIndex] <= process.MaxGasLimitPerMiniBlock
+	newGasSpentSender := mbc.gasSpentSender[dstIndex] + mbc.gasSpentSender[srcIndex]
+	newGasSpentReceiver := mbc.gasSpentReceiver[dstIndex] + mbc.gasSpentReceiver[srcIndex]
+	senderGasLimitOK := newGasSpentSender <= mbc.economicsFee.MaxGasLimitPerBlock()
+	receiverGasLimitOK := newGasSpentReceiver <= mbc.economicsFee.MaxGasLimitPerBlock()
+
+	canBeFullyMerged := senderGasLimitOK && receiverGasLimitOK
 	if canBeFullyMerged {
 		dstMiniBlock.TxHashes = append(dstMiniBlock.TxHashes, srcMiniBlock.TxHashes...)
-		mbc.mapMiniBlockGasConsumed[dstIndex] += mbc.mapMiniBlockGasConsumed[srcIndex]
+		mbc.gasSpentSender[dstIndex] += mbc.gasSpentSender[srcIndex]
+		mbc.gasSpentReceiver[dstIndex] += mbc.gasSpentReceiver[srcIndex]
 		return true
 	}
 
+	mbGasSpentSender := mbc.gasSpentSender[dstIndex]
+	mbGasSpentReceiver := mbc.gasSpentReceiver[dstIndex]
+
+	txHashes := mbc.getMaxTxsForMerge(&mbGasSpentSender, &mbGasSpentReceiver, srcMiniBlock)
+	if len(txHashes) > 0 {
+		dstMiniBlock.TxHashes = append(dstMiniBlock.TxHashes, txHashes...)
+		srcMiniBlock.TxHashes = mbc.removeTxHashesFromMiniBlock(srcMiniBlock, txHashes)
+
+		gasTransferred := mbGasSpentSender - mbc.gasSpentSender[dstIndex]
+		mbc.gasSpentSender[dstIndex] += gasTransferred
+		mbc.gasSpentSender[srcIndex] -= gasTransferred
+
+		gasTransferred = mbGasSpentReceiver - mbc.gasSpentReceiver[dstIndex]
+		mbc.gasSpentReceiver[dstIndex] += gasTransferred
+		mbc.gasSpentReceiver[srcIndex] -= gasTransferred
+	}
+
+	return false
+}
+
+func (mbc *miniBlocksCompaction) getMaxTxsForMerge(
+	gasSpentSender *uint64,
```

Comment:
maybe return these.

Reply:
These variables are modified inside this method and than used in the caller side